pkg/profiler/cpu.go

```diff
@@ -0,0 +1,909 @@
+				Help:        "Number of missing profile stacks",
+				ConstLabels: map[string]string{"target": target.String()},
+			},
+			[]string{"type"},
+		),
+		missingPIDs: promauto.With(reg).NewCounter(
+			prometheus.CounterOpts{
+				Name:        "parca_agent_profiler_missing_pid_total",
+				Help:        "Number of missing PIDs",
+				ConstLabels: map[string]string{"target": target.String()},
+			},
+		),
+		failedStackUnwindingAttempts: promauto.With(reg).NewCounterVec(
+			prometheus.CounterOpts{
+				Name:        "parca_agent_profiler_failed_stack_unwinding_attempts_total",
+				Help:        "Number of failed stack unwinding attempts",
+				ConstLabels: map[string]string{"target": target.String()},
+			},
+			[]string{"type"},
+		),
+		ksymCacheHitRate: promauto.With(reg).NewCounterVec(
+			prometheus.CounterOpts{
+				Name:        "parca_agent_profiler_kernel_symbolizer_cache_total",
+				Help:        "Hit rate for the kernel symbolizer cache",
+				ConstLabels: map[string]string{"target": target.String()},
+			},
+			[]string{"type"},
+		),
+	}
+}
+
+type CPUProfiler struct {
+	logger log.Logger
+
+	mtx    *sync.RWMutex
+	cancel func()
+
+	pidMappingFileCache *maps.PIDMappingFileCache
+	perfCache           *perf.Cache
+	ksymCache           *ksym.Cache
+	objCache            objectfile.Cache
+
+	bpfMaps   *bpfMaps
+	byteOrder binary.ByteOrder
+
+	lastError                      error
+	lastSuccessfulProfileStartedAt time.Time
+	nextProfileStartedAt           time.Time
+
+	writeClient profilestorepb.ProfileStoreServiceClient
+	debugInfo   *debuginfo.DebugInfo
+
+	target            model.LabelSet
+	profilingDuration time.Duration
+
+	discoveryManager discovery.Manager
+
+	profileBufferPool sync.Pool
+
+	metrics *metrics
+
+	pidLabels func() map[int]model.LabelSet
+}
+
+func NewCPUProfiler(
+	logger log.Logger,
+	reg prometheus.Registerer,
+	ksymCache *ksym.Cache,
+	objCache objectfile.Cache,
+	writeClient profilestorepb.ProfileStoreServiceClient,
+	debugInfoClient debuginfo.Client,
+	target model.LabelSet,
+	profilingDuration time.Duration,
+	pidLabels func() map[int]model.LabelSet,
+) Profiler {
+	return &CPUProfiler{
+		logger:              log.With(logger, "labels", target.String()),
+		mtx:                 &sync.RWMutex{},
+		target:              target,
+		profilingDuration:   profilingDuration,
+		writeClient:         writeClient,
+		ksymCache:           ksymCache,
+		pidMappingFileCache: maps.NewPIDMappingFileCache(logger),
+		perfCache:           perf.NewPerfCache(logger),
+		objCache:            objCache,
+		debugInfo: debuginfo.New(
+			log.With(logger, "component", "debuginfo"),
+			debugInfoClient,
+		),
+		profileBufferPool: sync.Pool{
+			New: func() interface{} {
+				return bytes.NewBuffer(nil)
+			},
+		},
+		byteOrder: byteorder.GetHostByteOrder(),
+		metrics:   newMetrics(reg, target),
+		pidLabels: pidLabels,
+	}
+}
+
+func (p *CPUProfiler) Name() string {
+	return "bpf-cpu-cycles"
```

Comment:
```suggestion
	return "bpf-cpu-sampled-stack-trace"
```
?

Reply:
The reason for the name is that I think it's a good indicator of what kind of sampling we are doing. In the future, we might want to add other perf events, such as cache miss counters and others (e.g.`bpf-cpu-l3-miss`) and I think this naming scheme would be clearer